src/main/java/wooteco/subway/domain/Sections.java

```diff
@@ -0,0 +1,225 @@
+        List<T> result = new ArrayList<>();
+        Long key = firstKey;
+
+        while (values.containsKey(key)) {
+            result.add(values.get(key));
+            key = newKeyMapper.apply(key);
+        }
+
+        return result;
+    }
+
+    public Section add(Section newSection) {
+        validateAddable(newSection);
+
+        if (getUpDestination().equals(newSection.getDownStation())) {
+            values.add(0, newSection);
+            return newSection;
+        }
+
+        if (getDownDestination().equals(newSection.getUpStation())) {
+            values.add(newSection);
+            return newSection;
+        }
+
+        for (Section section : values) {
+            if (section.hasSameUpStation(newSection)) {
+                return insertAfterUpStation(newSection, section);
+            }
+
+            if (section.hasSameDownStation(newSection)) {
+                return insertBeforeDownStation(newSection, section);
+            }
+        }
+        throw new IllegalStateException("구간을 추가할 수 있는 상태가 아닙니다.");
+    }
+
+    private Section insertAfterUpStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), newSection.getDownStation(), section.getDownStation(),
+            getNewDistance(section, newSection));
+        values.set(index, newSection);
+        values.add(index + 1, sectionToModify);
+        return sectionToModify;
+    }
+
+    private int getNewDistance(Section section, Section newSection) {
+        int newDistance = section.getDistance() - newSection.getDistance();
+        if (newDistance <= 0) {
+            throw new IllegalArgumentException("추가하려는 구간의 거리가 기존 구간의 거리보다 작아야합니다.");
+        }
+        return newDistance;
+    }
+
+    private Section insertBeforeDownStation(Section newSection, Section section) {
+        int index = values.indexOf(section);
+        Section sectionToModify = new Section(section.getId(), section.getUpStation(), newSection.getUpStation(),
+            getNewDistance(section, newSection));
+        values.set(index, sectionToModify);
+        values.add(index + 1, newSection);
+        return sectionToModify;
+    }
+
+    private void validateAddable(Section section) {
+        List<Section> foundSections = findSectionsOverlapped(section);
+
+        if (isInvalid(section, foundSections)) {
+            throw new IllegalArgumentException("이미 포함된 두 역을 가진 Section 을 추가할 수 없습니다.");
+        }
+    }
+
+    private List<Section> findSectionsOverlapped(Section newSection) {
+        return values.stream()
+            .filter(section -> section.isOverlap(newSection))
+            .collect(Collectors.toList());
+    }
+
+    private boolean isInvalid(Section section, List<Section> foundSections) {
+        return hasAnySectionWithSameStations(foundSections, section) ||
+            isInvalidCountInMiddle(foundSections) || isSameWithDestinations(foundSections);
```

Comment:
```suggestion
        return hasAnySectionWithSameStations(foundSections, section) ||
            isInvalidCountInMiddle(foundSections) || 
            isSameWithDestinations(foundSections);
```
한 줄씩 띄어주면 더 읽기 편할 것 같아요 😄

Reply:
한 줄에 코드가 너무 길지 않으면 개행을 넣지 않으려 했었는데 던이 남겨주신 피드백을 보고 판정 메서드마다 줄을 나눠서 가독성을 비교해보았습니다. 판정 메서드별로 줄을 나눈 것이 더 빠르게 파악할 수 있다는게 느껴져서 반영했습니다!! 좋은 피드백 감사합니다. 👍🏻👍🏻👍🏻👍🏻